doc/decisions/arbitrary_metadata.md

```diff
@@ -19,6 +21,8 @@ Most file formats do not have support for that.
 
 ## Decision
 
+Do not store metadata in any namespace but spec.
```

Comment:
Are we sure that this works? I'm still not convinced it works (losslessly) for arrays.

Also what happens for e.g. `kdb set-meta user:/myarray mymeta 5`?

Reply:
I think it should work, I clarified the array case in 349343cb07a219752bddfe7091c23c3a68d2fd49

Did you have something else in mind where it would break?

Reply:
I don't have a concrete case, I'm just not entirely convinced we didn't miss an edge case.

Your clarification actually confused me even more...

> In other namespaces only store metadata that is handled by the `spec` plugin.

What? `spec` only handles metadata from the `spec:/` namespace. 

> for `array`, metadata might be stored in any namespace

The way I understand your proposal, `array` should **never** be stored anywhere but `spec:/`. And that's because in `spec:/` it has a different meaning.

In `spec:/` the metadata `array=` (empty value) means  "this is an array". If you give it a value e.g. `array=#4` it means "this is an array with default size X" (e.g. `#4` = size 5).

In any other namespace (at runtime) `array=` means "this is an empty array" and e.g. `array=#4` means "this is an array with max index `#4`".

The `spec:/` meaning of `array=` must obviously be stored. The other meaning should _never_ be stored and instead should be inferred from the stored data. Either by the storage plugin (e.g. JSON array) or by the `spec` plugin. At least that is what I got from the sentence: "[if] the storage plugin does not support arrays, then the metadata will be discarded".

I can live with this approach, but just in case you didn't notice, this means we loose some (possibly important) functionality compared to the current approach. Take this spec:

```ini
[myarray]
array=#0

[myarray/#/data]
default=42
```

An application can (now and with your proposal) read `/myarray/#0/data` and will get `42`. The difference is, currently you can do 
```
kdb set-meta user:/myarray array #1
```
and then the `spec` plugin will generate 
```
user:/myarray/#0/data=42
user:/myarray/#1/data=42
```
The second key is the important one, because of it `/myarray/#1/data` will result in `42`. With your proposal this doesn't work, because `user:/` won't store the array size. Again I'm not quite sure what your intention was, maybe in such a case a plugin with array support like JSON should store e.g. `{ "myarray": [ { "data": 42 }, { "data": 42 } ] }`, i.e. store the defaults explicitly. But what about a plugin without array support?

The thing is arrays are a special case, so there will be some solution here. But that is actually why I asked you about the general case:

> Also what happens for e.g. kdb set-meta user:/myarray mymeta 5?

You didn't answer the question, so I assume you don't really know either... The way I understand your proposal this wouldn't do anything, unless `mymeta` can somehow be generated by `spec`. But if `spec` didn't generate `mymeta=5` before the `kdb set-meta` and it does afterwards, then **something** has to be stored in `user:/`, which means metadata **is** stored in `user:/`, just not explicitly.

Reply:
Thank you, here is the update: 459e489ac

> I'm just not entirely convinced we didn't miss an edge case.

I added the assumption that such a problem would be fixed by the definition in the respective metadata.

> because user:/ won't store the array size

Only not explicitly, of course it infers the correct size and sets the metadata `array` correctly. I added the constraint.

> `kdb set-meta user:/myarray mymeta 5` You didn't answer the question

For me this question was answered with the very first sentence "Do not store arbitrary metadata in any namespace except spec.", so this call cannot do anything useful. I clarified now it should fail and not discard the metadata.